re #5448: check that clause target type is usable at modality

plt-amy opened the pull request:
Implements the check for #5448, which at current `master` only happens with the  `--cubical-compatible` flag, eagerly (during LHS checking).

Because of how `transpX` clauses are generated (they use the RHS type, though with the unification substitutions represented as paths, in a relevant argument), checking that the RHS type is usable at the required modality will reject exactly the same definitions that `--cubical-compatible` would. With the benefit that the user doesn't have to see internal Cubical Agda code.

Consider:

```agda
subst : {A : Set} (@0 P : A → Set) {α β : A} (p : α ≡ β) → P α → P β
subst P refl a = {! a  !}
```

The error you get with `--without-K` is as follows. When `--cubical` is used, the note mentions "used for computing transports" and "in Cubical Agda" more specifically, and `--cubical-compatible` has yet a third wording. 

```
This clause has target type P α which is not usable at modality @ω

Note: when --without-K is enabled, the target type must be usable
at the modality in which the function was defined, since it is used
in substitutions.

when checking the clause left hand side
subst P refl a
```

plt-amy committed:
re #5448: check that clause target type is usable at modality

plt-amy committed:
fixup: tweak error message for --cubical-compatible slightly

plt-amy commented:
I've optimistically put this in the 2.6.3 milestone, in the hopes that there's enough time to sneak it in. But don't let it hold anyone up.

plt-amy committed:
fixup: update error messages w/ --cubical-compatible wording

nad commented on src/full/Agda/TypeChecking/Rules/LHS.hs:
```diff
@@ -701,6 +705,12 @@ checkLeftHandSide call f ps a withSub' strippedPats =
         addContext delta $ do
           mapM_ noShadowingOfConstructors eqs
 
+          -- If we're working --without-K, then we have to check the
```
This comment is unfinished.

nad commented on src/full/Agda/TypeChecking/Rules/LHS/Problem.hs:
```diff
@@ -236,6 +236,8 @@ data LHSState a = LHSState
     --   be type-checked in irrelevant mode.
   , _lhsPartialSplit :: ![Maybe Int]
     -- ^ have we splitted with a PartialFocus?
+  , _lhsIndexedSplit :: Bool
```
Perhaps this field could be strict.

nad commented on test/Fail/Issue5448-1.err:
```diff
@@ -1,6 +1,10 @@
-when checking the definition of subst
+Issue5448-1.agda:8,1-15
+This clause has target type P x which is not usable at the required
+modality
```
I think it also makes sense to change how the default modality is printed, as you pointed out somewhere else.

nad commented on src/full/Agda/TypeChecking/Irrelevance.hs:
```diff
@@ -480,18 +482,53 @@ instance UsableModality a => UsableModality (Arg a) where
+        context
+          | cubical    = "in Cubical Agda,"
+          | compatible = "to maintain compatibility with Cubical Agda,"
+          | otherwise  = "when --without-K is enabled,"
```
I don't think I saw all kinds of error messages in the test cases. Please add a test case that uses `--without-K`.

nad commented on src/full/Agda/TypeChecking/Irrelevance.hs:
```diff
@@ -480,18 +482,53 @@ instance UsableModality a => UsableModality (Arg a) where
+
+        justification
+          | (cubical || compatible) = "used for computing transports."
+          | otherwise               = "used in substitutions."
```
I'm not sure what you mean by this formulation.

plt-amy replied:
Ah, I was trying to find a neutral formulation of "used for computing transports", since there's no `transp`ort primitive. But maybe we should stick with that in all three cases, then?

nad replied:
Perhaps you could add `(in Cubical Agda)`.

nad reviewed (dismissed):
Great!

The current text in the changelog that describes the new flag `--cubical-compatible` should also be updated, perhaps along with some text in the user manual.

plt-amy committed:
fixup: finish comment, always print @ω in errors

plt-amy committed:
fixup: test --without-K error message, too

plt-amy committed:
fixup: reword --without-K justification

plt-amy commented:
@Mergifyio rebase

plt-amy commented:
(need a CI run with new std-lib commit to make sure nothing is broken)

nad commented on test/Fail/Issue5448-4-noK.err:
```diff
@@ -0,0 +1,10 @@
+
+Note: when --without-K is enabled, the target type must be usable
+at the modality in which the function was defined, since it is used
+to compute substitutions in Cubical Agda.
```
I think we should avoid mentioning Cubical Agda in the error message for `--without-K`. When `--without-K` is used we do not actually "compute substitutions in Cubical Agda", so the text might be misleading. Note that I'm currently not aware of anything that would go wrong if we removed this restriction for `--without-K`. However, the proof sketch that we use to motivate the current implementation suggests something like this restriction, and the proof sketch is for a language with cubical features.

By the way, what does "the modality in which the function was defined" mean?

plt-amy replied:
It's the attribute attached to the function's name:
```agda
@0 subst : (A : Set) (@0 : A → Set) {x y : A} → x ≡ y → P x → P y
subst ... = {! P x has to be usable @0 !}
```

Anyway, I can't think of a good wording for the `Note:` in the `--without-K` case, and I don't want to skip the check entirely because it contradicts "`--cubical-compatible` = `--without-K` + internal cubical code generation", since `--cubical-compatible` would (by design, rather than by a bug) reject more programs than `--without-K`

plt-amy replied:
How about `because in univalent type theory, substitution is computed by cases on the predicate`?

nad replied:
> How about `because in univalent type theory, substitution is computed by cases on the predicate`?

That may or may not be true, depending on how "univalent type theory" is defined.

Perhaps we can just skip (parts of) the explanation when `--without-K` is used.

plt-amy replied:
Works for me.

andreasabel commented on test/Fail/Issue5468.err:
```diff
@@ -1,4 +1,9 @@
-primTransp (λ i → x i) φ
-(unbox [ x₁ ]) is not usable at the required modality .
+This clause has target type primTransp (λ i → x i) φ (unbox [ x₁ ])
+which is not usable at the required modality .
```
Maybe for relevance there should be a different printer used here.  These modalities also have long forms, see https://github.com/agda/agda/blob/4bd0e883499ae7627de05a9e1f016ad5b001da13/src/full/Agda/Syntax/Concrete/Attribute.hs#L64-L69.
So, in this case `@irrelevant` would be preferable over `.` (which I took for a punctuation mark).

You are not to blame here, rather the past (this problem already existed).  However, it would make sense to change the standard _stand-alone_ printing of `Relevance` to the long forms `@relevant`, `@irrelevant`, `@shape-irrelevant`.  In a Pi-type or telescope, the short forms ` `, `.`, `..` should still be used.

plt-amy replied:
Thoughts on using `Verbalize` from `Agda.TypeChecking.Errors` here?

plt-amy replied:
Nevermind, it's too verbose then. I'll change the printing in the suggested way.

andreasabel commented on test/Fail/Issue5448-1.err:
```diff
@@ -1,6 +1,10 @@
-when checking the definition of subst
+Issue5448-1.agda:8,1-15
+This clause has target type P x which is not usable at the required
+modality @ω
```
This sentence should end with a dot.

andreasabel commented on test/Fail/Issue5448-1.err:
```diff
@@ -1,6 +1,10 @@
+
+Note: to maintain compatibility with Cubical Agda, the target type
+must be usable at the modality in which the function was defined,
+since it is used for computing transports.
```
This sentence should maybe not end with a dot.

andreasabel reviewed (changes requested):
Thanks, awesome!
Great if you can change the printing of relevances in the error message.

plt-amy committed:
Print modalities in error messages verbosely

... so that the irrelevance modality isn't confusable for punctuation
anymore. Also fixes up punctuation in the error message

andreasabel reviewed (approved):
Wow, that looks awesome!!

plt-amy committed:
fixup: remove justification when --without-K
